lib/ssl/src/ssl_gen_statem.erl

```diff
@@ -596,16 +596,11 @@ connection({call, From}, negotiated_protocol,
 		    [{reply, From, {ok, SelectedProtocol}}]);
 connection({call, From},
            {close, {Pid, _Timeout}},
-           #state{connection_env = #connection_env{terminated = closed} = CEnv,
-                 protocol_specific = PS} = State) ->
-    {next_state, downgrade, State#state{connection_env =
-                                            CEnv#connection_env{terminated = true,
-                                                                downgrade = {Pid, From}},
-                                        protocol_specific = PS#{active_n_toggle => true,
-                                                                active_n => 1}
-                                       },
-     [{next_event, internal, ?ALERT_REC(?WARNING, ?CLOSE_NOTIFY)}]};
-connection({call, From},
+           #state{connection_env = #connection_env{socket_tls_closed = true} = CEnv} = State) ->
+    %%Close Alert already received from peer, "replay" in downgrade state.
+    {next_state, downgrade, State#state{connection_env = CEnv#connection_env{downgrade = {Pid, From}}},
+                                        [{next_event, internal, ?ALERT_REC(?WARNING, ?CLOSE_NOTIFY)}]};
```

Comment:
Should the tls_sender:downgrade/2 call go here as well or not?

Reply:
Humm ... good question I actually think it should! I made a new version.